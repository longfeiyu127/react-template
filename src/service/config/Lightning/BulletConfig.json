{
  "player": {
    "size": {
      "height": 0.14,
      "width": 0.14
    },
    "position": {
      "top": -0.45,
      "left": 2
    },
    "targetPosition": {
      "top": "max",
      "left": "remain"
    },
    "bulletTpey": "player",
    "HP": 1,
    "speed": 0.02
  }
}