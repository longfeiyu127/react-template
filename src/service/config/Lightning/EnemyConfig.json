{
  "enemyA0": {
    "size": {
      "height": 0.45,
      "width": 0.45
    },
    "position": {
      "top": -0.45,
      "left": 2
    },
    "targetPosition": {
      "top": "max",
      "left": "remain"
    },
    "enemyType": "enemyA0",
    "HP": 1,
    "speed": 0.02,
    "status": "healthy"
  },
  "enemyA1": {
    "size": {
      "height": 0.45,
      "width": 0.45
    },
    "position": {
      "top": -0.45,
      "left": 2
    },
    "targetPosition": {
      "top": "max",
      "left": "remain"
    },
    "enemyType": "enemyA1",
    "HP": 1,
    "speed": 0.02,
    "status": "healthy"
  },
  "enemyA2": {
    "size": {
      "height": 0.45,
      "width": 0.45
    },
    "position": {
      "top": -0.45,
      "left": 2
    },
    "targetPosition": {
      "top": "max",
      "left": "remain"
    },
    "enemyType": "enemyA2",
    "HP": 1,
    "speed": 0.02,
    "status": "healthy"
  },
  "enemyC": {
    "size": {
      "height": 0.7,
      "width": 0.7
    },
    "position": {
      "top": -0.8,
      "left": 1.5
    },
    "targetPosition": {
      "top": "max",
      "left": "remain"
    },
    "enemyType": "enemyC",
    "HP": 3,
    "speed": 0.015,
    "status": "healthy"
  }
}