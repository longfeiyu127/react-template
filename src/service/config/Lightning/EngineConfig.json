{
  "step": 0.05,
  "bulletTpey": {
    "player": {
      "size": {},
      "position": {}
    }
  },
  "enemyType": {
    "enemyA0": {
      "size": {
        "height": 0,
        "width": 2
      },
      "position": {
        "top": 0,
        "left": 2
      },
      "targetPosition": {
        "top": 5,
        "left": 2
      },
      "HP": "",
      "speed": 0.01
    }
  }
}