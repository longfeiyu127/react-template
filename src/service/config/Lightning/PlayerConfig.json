{
  "size" : {
    "height": 0.56,
    "width": 0.8
  }
}