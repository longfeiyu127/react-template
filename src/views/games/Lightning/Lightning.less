.g-lightning{
  height: 100%;
  position: relative;
}