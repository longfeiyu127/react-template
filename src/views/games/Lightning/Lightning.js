import React from 'react';
import { Route } from 'react-router-dom';
// import Tictactoe from './tictactoe'
import Engine from '../../../components/games/Lightning/Engine/Engine';
import './Lightning.less';

class Lightning extends React.Component {
  constructor() {
    super();
    this.state = {
      // 选择难度
    };
  }

  render() {
    return (
      <div className="g-lightning">
        {/* <Route exact path={`${this.props.match.url}/tictactoe`} component={Tictactoe}/> */}
        {/* <Route exact path={`${this.props.match.url}/fivechess`} component={Fivechess}/> */}
        {/* <Route exact path={`${this.props.match.url}/games`} component={Games}/>
        <Route exact path={`${this.props.match.url}/ranks`} component={Ranks}/>
        <Route exact path={`${this.props.match.url}/developer`} component={Developer}/> */}
        <Route exact path="/games/lightning" component={Engine} />
      </div>
    );
  }
}

export default Lightning;
