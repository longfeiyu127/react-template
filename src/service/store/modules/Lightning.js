import EnemyConfig from '../../config/Lightning/EnemyConfig';
import BulletConfig from '../../config/Lightning/BulletConfig';
// import EngineConfig from '../../../../service/config/Lightning/EngineConfig';
import PlayerConfig from '../../config/Lightning/PlayerConfig';

// const config = {
//   step: 0.05, // 控住步长
//   bulletTpey: {
//     // 玩家子弹
//     player: {
//       size: {},
//       position: {}
//     }
//     // 玩家导弹
//     // 敌机子弹
//   },
//   enemyType: {
//     enemyA0: {
//       size: {
//         height: 0,
//         width: 2
//       },
//       position: {
//         top: 0,
//         left: 2
//       },
//       targetPosition: {
//         top: 5,
//         left: 2
//       },
//       HP: '',
//       speed: 0.01
//     }
//   }
// };

function deepCopy(params) {
  return JSON.parse(JSON.stringify(params));
}

// 移动敌机
function EnemyPosition(enemy) {
  const newEnemy = deepCopy(enemy);
  newEnemy.position.top = -0.5 * enemy.size.height;
  newEnemy.position.left = Math.random() * (3.75 - enemy.size.width);
  newEnemy.targetPosition.left = enemy.position.left;
  return newEnemy;
}

// 创建敌机
function createEnemy(state) {
  const { EnemyArr } = state;
  const newEnemy = [];
  const EnemyWeightA = Math.random() * 100;
  if (EnemyWeightA < 10) {
    newEnemy.push(EnemyPosition(EnemyConfig.enemyA0));
  } else if (EnemyWeightA >= 10 && EnemyWeightA < 20) {
    newEnemy.push(EnemyPosition(EnemyConfig.enemyA1));
  } else if (EnemyWeightA >= 30 && EnemyWeightA < 40) {
    newEnemy.push(EnemyPosition(EnemyConfig.enemyA2));
  }
  const EnemyWeightC = Math.random() * 100;
  if (EnemyWeightC < 2) {
    newEnemy.push(EnemyPosition(EnemyConfig.enemyC));
  }
  return EnemyArr.concat(newEnemy);
}

// 创建子弹
function createBullet(state) {
  const { BulletArr, playerTop, playerLeft } = state;
  const position = {
    top: playerTop,
    left:
      playerLeft +
      PlayerConfig.size.width * 0.5 -
      BulletConfig.player.size.width * 0.5
  };
  const targetPosition = {
    top: 'min',
    left: 0
    // left: this.playerLeft
  };
  const newBullet = {};
  Object.assign(newBullet, BulletConfig.player);
  newBullet.position = position;
  newBullet.targetPosition = targetPosition;
  BulletArr.push(newBullet);
  return BulletArr;
}

// 处理敌机移动
function EnemyControll(EnemyArr, screen) {
  const EnemyArrRes = EnemyArr.map(item => {
    // console.log(config.enemyType[item.enemyType])
    // console.log(item.speed);
    // eslint-disable-next-line
    item.position.top += item.speed;
    return item;
  });
  return EnemyArrRes.filter(
    item =>
      // item.position.top <= (screen.clientHeight / screen.clientWidth) * 3.75
      item.position.top <= screen.clientHeight / 100
  );
}

//  处理子弹移动(处理超出边界子弹)
function BulletControll(BulletArr) {
  const BulletArrRes = BulletArr.map(item => {
    // eslint-disable-next-line
    item.position.top -= item.speed;
    return item;
  });
  return BulletArrRes.filter(item => item.position.top >= 0);
}

// 检查碰撞
function crashEngine(EnemyArr, BulletArr) {
  const RemainsArr = [];
  EnemyArr.map((EnemyItem, EnemyIndex) => {
    BulletArr.map((BulletItem, BulletIndex) => {
      // console.log(EnemyItem, BulletItem)
      // const { EnemyX, EnemyY, BulletX, BulletY } = {}
      const EnemyX = EnemyItem.position.left + EnemyItem.size.width * 0.5;
      const EnemyY = EnemyItem.position.top + EnemyItem.size.height * 0.5;
      const BulletX = BulletItem.position.left + BulletItem.size.width * 0.5;
      const BulletY = BulletItem.position.top + BulletItem.size.height * 0.5;
      const CenterDistance =
        (EnemyItem.size.width + BulletItem.size.width) * 0.5;
      if (
        Math.sqrt((BulletX - EnemyX) ** 2 + (BulletY - EnemyY) ** 2) <
        CenterDistance
      ) {
        const newRemains = { ...EnemyItem };
        RemainsArr.push(newRemains);
        EnemyArr.splice(EnemyIndex, 1);
        BulletArr.splice(BulletIndex, 1);
        // console.log('pengdaole')
        // console.log(EnemyX, EnemyY)
        // console.log(BulletX, BulletY)
        // console.log(EnemyItem.size.width*0.5, BulletItem.size.width*0.5)
        // console.log(Math.sqrt(Math.pow(BulletX-EnemyX, 2) + Math.pow(BulletY - EnemyY, 2)))
        // console.log((EnemyItem.size.width + BulletItem.size.width))
        // alert('pengdaole')
      }
      return null;
    });
    return null;
  });
  return {
    newRemainsArr: RemainsArr,
    newEnemyArr: EnemyArr,
    newBulletArr: BulletArr
  };
}

const Lightning = {
  state: {
    // 运算状态
    EngineStatus: false,
    Enginespeed: 15,
    // Enginespeed: 100,
    // 玩家位置
    playerTop: 3.8,
    playerLeft: 1.35,
    // 子弹数组
    BulletArr: [],
    // 敌机数组
    EnemyArr: [],
    // 残骸数组
    RemainsArr: []
  },
  reducers: {
    setEngineStatus(state, bol) {
      return {
        ...state,
        EngineStatus: bol
      };
    },
    // 创建引擎
    creator(state) {
      const EnemyArr = createEnemy(deepCopy(state));
      const BulletArr = createBullet(deepCopy(state));
      return {
        ...state,
        EnemyArr,
        BulletArr
      };
    },
    // 移动位置/检查碰撞
    mainEngine(state, screen) {
      const { EngineStatus } = state;
      if (EngineStatus) {
        // 控制移动
        const EnemyArr = EnemyControll(deepCopy(state.EnemyArr), screen);
        const BulletArr = BulletControll(deepCopy(state.BulletArr));
        // 检查碰撞
        const { newRemainsArr, newEnemyArr, newBulletArr } = crashEngine(
          EnemyArr,
          BulletArr
        );
        // 处理完成，设置数据
        return {
          ...state,
          RemainsArr: state.RemainsArr.concat(newRemainsArr),
          BulletArr: newBulletArr,
          EnemyArr: newEnemyArr
        };
      }
      return {
        ...state
      };
    },
    // 删除残骸
    delRemains(state, index) {
      return {
        ...state,
        RemainsArr: state.RemainsArr.splice(index, 1)
      };
    },
    // 控制主机移动
    ControllPlayer(state, positionChange) {
      // console.log(positionChange);
      const { playerTop, playerLeft } = state;
      return {
        ...state,
        playerTop: parseInt((playerTop + positionChange.top) * 100, 10) / 100,
        playerLeft: parseInt((playerLeft + positionChange.left) * 100, 10) / 100
      };
    }
  },
  effects: {}
};

export default Lightning;
