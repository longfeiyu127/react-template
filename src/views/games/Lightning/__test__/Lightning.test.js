import React from 'react';
import renderer from 'react-test-renderer';
import { MemoryRouter } from 'react-router';
import Lightning from '../Lightning';

describe('<Lightning />', () => {
  it('component render', () => {
    let lightning = renderer
      .create(
        <MemoryRouter>
          <Lightning />
        </MemoryRouter>
      )
      .toJSON();
    expect(lightning).toMatchSnapshot();

    lightning = renderer
      .create(
        <MemoryRouter>
          <Lightning />
        </MemoryRouter>
      )
      .toJSON();
    expect(lightning).toMatchSnapshot();
  });
});
