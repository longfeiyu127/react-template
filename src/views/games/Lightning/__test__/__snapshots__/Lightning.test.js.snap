// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Lightning /> component render 1`] = `
<div
  className="g-lightning"
/>
`;

exports[`<Lightning /> component render 2`] = `
<div
  className="g-lightning"
/>
`;
